src/tqec/computation/open_graph.py

```diff
@@ -0,0 +1,230 @@
+from tqec.computation.cube import YHalfCube, ZXCube
+from tqec.utils.enums import Basis
+from tqec.utils.exceptions import TQECException
+from tqec.utils.position import Direction3D
+
+
+@dataclass(frozen=True)
+class FilledGraph:
+    """A block graph without open ports that can be used for simulation.
+
+    Attributes:
+        graph: The block graph with all ports filled.
+        stabilizers: The external stabilizers simulable on the filled graph.
+        observables: The correlation surfaces that can be used as logical observables
+            for the simulation. A list of independent correlation surfaces with the
+            area as small as possible is chosen.
+    """
+
+    graph: BlockGraph
+    stabilizers: list[str]
+    observables: list[CorrelationSurface]
+
+    def __post_init__(self) -> None:
+        if self.graph.num_ports != 0:
+            raise TQECException("The filled graph should not have open ports.")
+        if len(self.stabilizers) != len(self.observables):
+            raise TQECException(
+                "The number of stabilizers and observables should match."
+            )
+
+    def get_external_stabilizers(self) -> list[str]:
+        """Return the external stabilizers of the correlation surfaces."""
+        return [
+            obs.external_stabilizer_on_graph(self.graph) for obs in self.observables
+        ]
+
+
+def fill_ports_for_minimal_simulation(graph: BlockGraph) -> list[FilledGraph]:
+    """Given a block graph with open ports, fill in the ports with the appropriate
+    cubes that will minimize the number of simulation runs needed for the complete
+    logical observable set.
+
+    Args:
+        graph: The block graph with open ports.
+
+    Returns:
+        A list of :class:`~tqec.computation.open_graph.FilledGraph` instances, each
+        containing a block graph with all ports filled and a set of correlation
+        surfaces that can be used as logical observables for the simulation on that
+        block graph.
+    """
+    num_ports = graph.num_ports
+    if num_ports == 0:
+        raise TQECException("The provided graph has no open ports.")
+    correlation_surfaces = graph.find_correlation_surfaces()
+    ext_stabs = [s.external_stabilizer_on_graph(graph) for s in correlation_surfaces]
+    stab_to_surface: dict[str, CorrelationSurface] = {
+        stab: surface for stab, surface in zip(ext_stabs, correlation_surfaces)
+    }
+    identity = "I" * num_ports
+    # Need to collect all the possible correlation surfaces because we want
+    # to find the generators with the smallest correlation surface area
+    for stabilizer, comb in _iter_stabilizer_group(ext_stabs):
+        if stabilizer != identity and stabilizer not in stab_to_surface:
+            correlation_surface = reduce(
+                lambda a, b: a ^ b,
+                [stab_to_surface[s] for s in comb],
+            )
+            stab_to_surface[stabilizer] = correlation_surface
```

Comment:
Here, is it guaranteed that if `stabilizer in stab_to_surface` then the correlation_surface associated to the stabilizer is the same? Basically, I am wondering if the following code

```suggestion
        if stabilizer != identity:
            correlation_surface = reduce(
                lambda a, b: a ^ b,
                [stab_to_surface[s] for s in comb],
            )
            stab_to_surface[stabilizer] = (
                correlation_surface
                if stabilizer not in stab_to_surface else
                min(
                    (correlation_surface, stab_to_surface[stabilizer]), 
                    key=lambda cs: cs.area()
                )
            )
```

should be rather considered or if we know for sure that the computed `correlation_surface` will always be equal to `stab_to_surface[stabilizer]`.

Reply:
There are two cases to consider:

1. If the initial generating set of the correlation surface from `find_correlation_surface` is exact-complete, the unique decomposition of stabilizers into these generators ensures that no stabilizer will be repeated during iteration through the group.
2. If the generating set is over-complete and a stabilizer has two different correlation surfaces => the identity stabilizer could have a non-empty correlation surface. That is possible but leads to an unusual structure, such as loops in the diagram.

Since we do not guarantee that the generators are exact-complete, you are right—we need to include this check. However, I would prefer to first reduce the generating set to an exact-complete form and then iterate over the group generated by them. This approach not only ensures correctness but also reduces the total number of iterations.

Reply:
> I would prefer to first reduce the generating set to an exact-complete form and then iterate over the group generated by them

This has been addressed in the last commit.